File: baseLib/src/main/java/me/ycdev/android/lib/common/utils/WeakListenerManager.java
package me.ycdev.android.lib.common.utils;

import android.support.annotation.NonNull;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

public class WeakListenerManager<IListener> {
    private static final String TAG = "WeakListenerManager";

    public interface NotifyAction<IListener> {
        void notify(IListener listener);
    }

    private class ListenerInfo {
        String className;
        WeakReference<IListener> holder;

        ListenerInfo(IListener listener) {
            className = listener.getClass().getName();
            holder = new WeakReference<>(listener);
        }
    }

    private final List<ListenerInfo> mListeners = new ArrayList<>();

    /**
     * Only invoked when invoke {@link #addListener(Object)}
     */
    protected void onFirstListenerAdd() {
        // nothing to do
    }

    /**
     * Only invoked when invoke {@link #removeListener(Object)}
     */
    protected void onLastListenerRemoved() {
        // nothing to do
    }

    /**
     * Override this method to notify the listener when registered.
     */
    protected void onListenerAdded(@NonNull IListener listener) {
        // nothing to do
    }

    public void addListener(@NonNull IListener listener) {
        synchronized (mListeners) {
            if (mListeners.size() == 0) {
                onFirstListenerAdd();
            }

            for (ListenerInfo l : mListeners) {
                if (l.holder.get() == listener) return; // skip duplicate listeners
            }
            mListeners.add(new ListenerInfo(listener));
        }

        // Notify the listener to get initialized
        onListenerAdded(listener);
    }

    public void removeListener(@NonNull IListener listener) {
        synchronized (mListeners) {
            final int N = mListeners.size();
            boolean removed = false;
            for (int i = 0; i < N; i++) {
                ListenerInfo listenerInfo = mListeners.get(i);
                if (listenerInfo.holder.get() == listener) {
                    mListeners.remove(i);
                    removed = true;
                    break;
                }
            }
            if (mListeners.size() == 0 && removed) {
                onLastListenerRemoved();
            }
        }
    }

    public void notifyListeners(@NonNull NotifyAction<IListener> action) {
        synchronized (mListeners) {
            for (int i = 0; i < mListeners.size();) {
                ListenerInfo listenerInfo = mListeners.get(i);
                IListener l = listenerInfo.holder.get();
                if (l == null) {
                    LibLogger.e(TAG, "listener leak found: " + listenerInfo.className);
                    mListeners.remove(i);
                } else {
                    LibLogger.d(TAG, "notify: " + listenerInfo.className);
                    action.notify(l);
                    i++;
                }
            }
            LibLogger.d(TAG, "notify done, cur size: " + mListeners.size());
        }
    }
}

File: baseLib/src/main/java/me/ycdev/android/lib/common/utils/WeakListenerManager.kt
package me.ycdev.android.lib.common.utils

import androidx.annotation.VisibleForTesting
import timber.log.Timber
import java.lang.ref.WeakReference
import java.util.ArrayList

@Suppress("unused")
open class WeakListenerManager<IListener : Any> {
    private val listeners = ArrayList<ListenerInfo<IListener>>()

    interface NotifyAction<IListener> {
        fun notify(listener: IListener)
    }

    private class ListenerInfo<IListener : Any>(listener: IListener) {
        var className: String = listener::class.java.name
        var holder: WeakReference<IListener> = WeakReference(listener)
    }

    /**
     * Only invoked when invoke [.addListener]
     */
    protected open fun onFirstListenerAdd() {
        // nothing to do
    }

    /**
     * Only invoked when invoke [.removeListener]
     */
    protected open fun onLastListenerRemoved() {
        // nothing to do
    }

    /**
     * Override this method to notify the listener when registered.
     */
    protected open fun onListenerAdded(listener: IListener) {
        // nothing to do
    }

    fun addListener(listener: IListener) {
        synchronized(listeners) {
            if (listeners.size == 0) {
                onFirstListenerAdd()
            }

            for (l in listeners) {
                if (l.holder.get() === listener) return // skip duplicate listeners
            }
            listeners.add(ListenerInfo(listener))
        }

        // Notify the listener to get initialized
        onListenerAdded(listener)
    }

    fun removeListener(listener: IListener) {
        synchronized(listeners) {
            var removed = false
            for (i in 0 until listeners.size) {
                val listenerInfo = listeners[i]
                if (listenerInfo.holder.get() === listener) {
                    listeners.removeAt(i)
                    removed = true
                    break
                }
            }
            if (listeners.size == 0 && removed) {
                onLastListenerRemoved()
            }
        }
    }

    fun notifyListeners(action: NotifyAction<IListener>) {
        notifyListeners { action.notify(it) }
    }

    fun notifyListeners(action: (IListener) -> Unit) {
        synchronized(listeners) {
            // The listener may unregister itself!
            val listenersCopied: List<ListenerInfo<IListener>> = ArrayList(listeners)
            for ((i, listenerInfo) in listenersCopied.withIndex()) {
                val l = listenerInfo.holder.get()
                if (l == null) {
                    Timber.tag(TAG).w("listener leak found: %s", listenerInfo.className)
                    listeners.remove(listenerInfo)
                } else {
                    Timber.tag(TAG).d("notify #%d: %s", i, listenerInfo.className)
                    action(l)
                }
            }
            Timber.tag(TAG).d("notify done, cur size: %d in %s", listeners.size, this)
        }
    }

    @VisibleForTesting
    internal fun listenersCount(): Int {
        return listeners.size
    }

    companion object {
        private const val TAG = "WeakListenerManager"
    }
}
